File: 3. Behavioral Design Patterns/Mediator/Go/MediatorTemplate.go
package main

// Mediator Pattern – Go Template

type Mediator interface {
	Notify(sender interface{}, eventType string)
}

type Colleague interface {
	DoSomething()
}

type ConcreteColleagueA struct {
	mediator Mediator
}

func NewConcreteColleagueA(m Mediator) *ConcreteColleagueA {
	return &ConcreteColleagueA{mediator: m}
}

func (c *ConcreteColleagueA) DoSomething() {
	c.mediator.Notify(c, "EventA")
}

type ConcreteColleagueB struct {
	mediator Mediator
}

func NewConcreteColleagueB(m Mediator) *ConcreteColleagueB {
	return &ConcreteColleagueB{mediator: m}
}

func (c *ConcreteColleagueB) DoSomething() {
	c.mediator.Notify(c, "EventB")
}

type ConcreteMediator struct {
	colleagues []Colleague
}

func NewConcreteMediator() *ConcreteMediator {
	return &ConcreteMediator{colleagues: make([]Colleague, 0)}
}

func (m *ConcreteMediator) AddColleague(colleague Colleague) {
	m.colleagues = append(m.colleagues, colleague)
}

func (m *ConcreteMediator) Notify(sender interface{}, eventType string) {
	/* handle event */
}

func main() {
	mediator := NewConcreteMediator()
	colleagueA := NewConcreteColleagueA(mediator)
	colleagueB := NewConcreteColleagueB(mediator)
	mediator.AddColleague(colleagueA)
	mediator.AddColleague(colleagueB)
	colleagueA.DoSomething()
}

File: 3. Behavioral Design Patterns/Mediator/Python/MediatorTemplate.py
# Mediator Pattern – Python Template

from abc import ABC, abstractmethod

class Mediator(ABC):
    @abstractmethod
    def notify(self, sender, event_type):
        pass

class Colleague(ABC):
    def __init__(self, mediator):
        self.mediator = mediator

class ConcreteColleagueA(Colleague):
    def do_something(self):
        self.mediator.notify(self, "EventA")

class ConcreteColleagueB(Colleague):
    def do_something(self):
        self.mediator.notify(self, "EventB")

class ConcreteMediator(Mediator):
    def __init__(self):
        self.colleagues = []
    
    def add_colleague(self, colleague):
        self.colleagues.append(colleague)
    
    def notify(self, sender, event_type):
        print(f"Mediator handling {event_type} from {type(sender).__name__}")

def main():
    mediator = ConcreteMediator()
    colleague_a = ConcreteColleagueA(mediator)
    colleague_b = ConcreteColleagueB(mediator)
    mediator.add_colleague(colleague_a)
    mediator.add_colleague(colleague_b)
    colleague_a.do_something()

if __name__ == "__main__":
    main()
